python/ray/ml/train/integrations/sklearn/sklearn_trainer.py

```diff
@@ -112,26 +157,273 @@ def __init__(
+            return_train_score=self.return_train_score_cv,
         )
+        cv_aggregates = {}
+        for k, v in cv_results.items():
+            if not isinstance(v, np.ndarray):
+                continue
+            try:
+                cv_aggregates[f"{k}_mean"] = np.nanmean(v)
+                cv_aggregates[f"{k}_std"] = np.nanstd(v)
+            except Exception as e:
+                logger.warning(
+                    f"Couldn't calculate aggregate metrics for CV folds! {e}"
+                )
+                cv_aggregates[f"{k}_mean"] = np.nan
+                cv_aggregates[f"{k}_std"] = np.nan
+                pass
+
+        return {"cv": {**cv_results, **cv_aggregates}}
+
+    def _get_cv_parallelism(self, has_gpus: bool) -> bool:
+        parallelize_cv = False
+
+        assert not (has_gpus and self.parallelize_cv)
+
+        estimator_has_parallelism_params = has_cpu_params(self.estimator)
+
+        if self.cv and self.parallelize_cv is True:
+            parallelize_cv = True
+
+        if (
+            not has_gpus
+            and self.cv
+            and self.parallelize_cv is None
+            and not estimator_has_parallelism_params
+        ):
+            logger.info(
+                "No parallelism-related params detected in estimator, "
+                "will parallelize cross-validation instead."
+            )
+            parallelize_cv = True
+        return parallelize_cv
 
     def training_loop(self) -> None:
-        return super().training_loop()
+        register_ray()
+
+        self.estimator.set_params(**self.params)
+
+        datasets = self._get_datasets()
+        X_train, y_train = datasets.pop(TRAIN_DATASET_KEY)
+        groups = None
+        if "cv_groups" in X_train.columns:
+            groups = X_train["cv_groups"]
+            X_train = X_train.drop("cv_groups", axis=1)
+
+        trial_resources = tune.get_trial_resources().required_resources
+        num_cpus = int(trial_resources.get("CPU", 0))
+        has_gpus = bool(trial_resources.get("GPU", 0))
+
+        # see https://scikit-learn.org/stable/computing/parallelism.html
+        os.environ["OMP_NUM_THREADS"] = str(num_cpus)
+        os.environ["MKL_NUM_THREADS"] = str(num_cpus)
+        os.environ["OPENBLAS_NUM_THREADS"] = str(num_cpus)
+        os.environ["BLIS_NUM_THREADS"] = str(num_cpus)
+
+        parallelize_cv = self._get_cv_parallelism(has_gpus)
+        if self.set_estimator_cpus:
+            num_estimator_cpus = 1 if parallelize_cv else num_cpus
+            set_cpu_params(self.estimator, num_estimator_cpus)
+
+        with parallel_backend("ray", n_jobs=num_cpus):
+            start_time = time()
+            self.estimator.fit(X_train, y_train, **self.fit_params)
+            fit_time = time() - start_time
+
+            with tune.checkpoint_dir(step=1) as checkpoint_dir:
+                with open(os.path.join(checkpoint_dir, MODEL_KEY), "wb") as f:
+                    cpickle.dump(self.estimator, f)
```

Comment:
Can we add the preprocessor directly here instead of in `_postprocess_checkpoint`? We should work towards removing `_postprocess_checkpoint` I feel.

Reply:
Yes, we should be able to do it here